File: CS/BO.cs
using System;
using System.Collections.Generic;
using System.Text;
using DevExpress.Data;
using System.Reflection;
using System.Collections;

namespace WindowsApplication1 {

    #region CustomAttribute
    [AttributeUsage(AttributeTargets.All, AllowMultiple = true)]

    public class Reportable : Attribute {
        private string altName;
        public string AlternateName {
            get { return this.altName; }
            set { this.altName = value; }
        }
    }
    
    #endregion
    #region BO - Phone class
    public class Phone {
        string phoneNumber;
        [Reportable(AlternateName = "Phone number")]
        public string PhoneNumber {
            get { return phoneNumber; }
            set { phoneNumber = value; }
        }
        string contanctName;
        [Reportable(AlternateName = "Contact Name")]
        public string ContanctName {
            get { return contanctName; }
            set { contanctName = value; }
        }
        public Phone(string number, string name) {
            contanctName = name;
            phoneNumber = number;
        }
    } 
    #endregion
    #region BO - Address
    public class Address {
        string city;
        string business_address;
        string secondaryCode;
        public Address() { phones = new List<Phone>(); }
        public Address(string City, string Address, string SecondaryCode) {
            city = City;
            business_address = Address;
            secondaryCode = SecondaryCode;
            phones = new List<Phone>();
        }
        public string SecondaryCode {
            get { return secondaryCode; }
            set { secondaryCode = value; }
        }

        [Reportable()]
        public string City {
            get { return city; }
            set { city = value; }
        }
        [Reportable(AlternateName = "Business address")]
        public string Business_address {
            get { return business_address; }
            set { business_address = value; }
        }

        List<Phone> phones;
        [Reportable(AlternateName = "Public phones")]
        public List<Phone> Phones {
            get { return phones; }
        }
        public void AddPhone(Phone phone) {
            phones.Add(phone);
        }
    } 
    #endregion
    #region BO - Company
    [Reportable()]
    public class Company {
        public Company() { addressList = new List<Address>(); }
        public Company(string CompanyName, string PublicEMail, string RegistrationID) {
            name = CompanyName;
            emailAddress = PublicEMail;
            registrationID = RegistrationID;
            addressList = new List<Address>();
        }
        public void AddAddress(Address address) {
            addressList.Add(address);
        }

        string name;
        string emailAddress;
        string registrationID;
        List<Address> addressList;

        [Reportable(AlternateName = "Company Name")]
        public string Name {
            get { return name; }
            set { name = value; }
        }

        [Reportable(AlternateName = "e-mail Address")]
        public string EmailAddress {
            get { return emailAddress; }
            set { emailAddress = value; }
        }

        [Reportable(AlternateName = "Registration Number")]
        public string RegistrationID {
            get { return registrationID; }
            set { registrationID = value; }
        }

        [Reportable(AlternateName = "Address")]
        public List<Address> RptOnlyAddress {
            get {
                return addressList;
            }
            set {
                addressList = value;
            }
        }

    } 
    #endregion
    #region IDataDictionary Implementation - Companies collection
    public class Companies : List<Company>, IDataDictionary {
        #region IDataDictionary Members

        public string GetDataSourceDisplayName() {
            return "Companies List";
        }

        public string GetObjectDisplayName(string dataMember) {
            if(dataMember != "" && this.Count > 0) {
                string[] names = dataMember.Split('.');
                string[] altNames = dataMember.Split('.');
                object mObj = this[0];
                Type mType = mObj.GetType();

                for(int i = 0; i < names.Length; i++) {
                    PropertyInfo pi = mType.GetProperty(names[i]);
                    if(pi != null) {
                        Reportable[] reportable = (Reportable[])pi.GetCustomAttributes(typeof(Reportable), false);
                        if(reportable.Length > 0) {
                            if(reportable[0].AlternateName != null)
                                altNames[i] = reportable[0].AlternateName;
                        }
                        else
                            return "";

                    }
                    else
                        return dataMember;
                    if (pi.PropertyType.IsGenericType && !pi.PropertyType.IsValueType) {
                        mObj = pi.GetValue(mObj, null);
                        mObj = ((IList)mObj)[0];
                        mType = mObj.GetType();
                    }

                }
                string s = String.Join(".", altNames);
                return s;
            }
            else
                return dataMember;
        }


        #endregion
    } 
    #endregion

}

File: VB/BO.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Text
Imports DevExpress.Data
Imports System.Reflection
Imports System.Collections

Namespace WindowsApplication1

	#Region "CustomAttribute"
	<AttributeUsage(AttributeTargets.All, AllowMultiple := True)> _
	Public Class Reportable
		Inherits Attribute
		Private altName As String
		Public Property AlternateName() As String
			Get
				Return Me.altName
			End Get
			Set
				Me.altName = Value
			End Set
		End Property
	End Class

	#End Region
	#Region "BO - Phone class"
	Public Class Phone
		Private phoneNumber_Renamed As String
		<Reportable(AlternateName := "Phone number")> _
		Public Property PhoneNumber() As String
			Get
				Return phoneNumber_Renamed
			End Get
			Set
				phoneNumber_Renamed = Value
			End Set
		End Property
		Private contanctName_Renamed As String
		<Reportable(AlternateName := "Contact Name")> _
		Public Property ContanctName() As String
			Get
				Return contanctName_Renamed
			End Get
			Set
				contanctName_Renamed = Value
			End Set
		End Property
		Public Sub New(ByVal number As String, ByVal name As String)
			contanctName_Renamed = name
			phoneNumber_Renamed = number
		End Sub
	End Class
	#End Region
	#Region "BO - Address"
	Public Class Address
		Private city_Renamed As String
		Private business_address_Renamed As String
		Private secondaryCode_Renamed As String
		Public Sub New()
			phones_Renamed = New List(Of Phone)()
		End Sub
		Public Sub New(ByVal City As String, ByVal Address As String, ByVal SecondaryCode As String)
			Me.city_Renamed = City
			business_address_Renamed = Address
			Me.secondaryCode_Renamed = SecondaryCode
			phones_Renamed = New List(Of Phone)()
		End Sub
		Public Property SecondaryCode() As String
			Get
				Return secondaryCode_Renamed
			End Get
			Set
				secondaryCode_Renamed = Value
			End Set
		End Property

		<Reportable()> _
		Public Property City() As String
			Get
				Return city_Renamed
			End Get
			Set
				city_Renamed = Value
			End Set
		End Property
		<Reportable(AlternateName := "Business address")> _
		Public Property Business_address() As String
			Get
				Return business_address_Renamed
			End Get
			Set
				business_address_Renamed = Value
			End Set
		End Property

		Private phones_Renamed As List(Of Phone)
		<Reportable(AlternateName := "Public phones")> _
		Public ReadOnly Property Phones() As List(Of Phone)
			Get
				Return phones_Renamed
			End Get
		End Property
		Public Sub AddPhone(ByVal phone As Phone)
			phones_Renamed.Add(phone)
		End Sub
	End Class
	#End Region
	#Region "BO - Company"
	<Reportable()> _
	Public Class Company
		Public Sub New()
			addressList = New List(Of Address)()
		End Sub
		Public Sub New(ByVal CompanyName As String, ByVal PublicEMail As String, ByVal RegistrationID As String)
			name_Renamed = CompanyName
			emailAddress_Renamed = PublicEMail
			Me.registrationID_Renamed = RegistrationID
			addressList = New List(Of Address)()
		End Sub
		Public Sub AddAddress(ByVal address As Address)
			addressList.Add(address)
		End Sub

		Private name_Renamed As String
		Private emailAddress_Renamed As String
		Private registrationID_Renamed As String
		Private addressList As List(Of Address)

		<Reportable(AlternateName := "Company Name")> _
		Public Property Name() As String
			Get
				Return name_Renamed
			End Get
			Set
				name_Renamed = Value
			End Set
		End Property

		<Reportable(AlternateName := "e-mail Address")> _
		Public Property EmailAddress() As String
			Get
				Return emailAddress_Renamed
			End Get
			Set
				emailAddress_Renamed = Value
			End Set
		End Property

		<Reportable(AlternateName := "Registration Number")> _
		Public Property RegistrationID() As String
			Get
				Return registrationID_Renamed
			End Get
			Set
				registrationID_Renamed = Value
			End Set
		End Property

		<Reportable(AlternateName := "Address")> _
		Public Property RptOnlyAddress() As List(Of Address)
			Get
				Return addressList
			End Get
			Set
				addressList = Value
			End Set
		End Property

	End Class
	#End Region
	#Region "IDataDictionary Implementation - Companies collection"
	Public Class Companies
		Inherits List(Of Company)
		Implements IDataDictionary
		#Region "IDataDictionary Members"

		Public Function GetDataSourceDisplayName() As String Implements IDataDictionary.GetDataSourceDisplayName
			Return "Companies List"
		End Function

		Public Function GetObjectDisplayName(ByVal dataMember As String) As String Implements IDataDictionary.GetObjectDisplayName
			If dataMember <> "" AndAlso Me.Count > 0 Then
				Dim names As String() = dataMember.Split("."c)
				Dim altNames As String() = dataMember.Split("."c)
				Dim mObj As Object = Me(0)
				Dim mType As Type = mObj.GetType()

				Dim i As Integer = 0
				Do While i < names.Length
					Dim pi As PropertyInfo = mType.GetProperty(names(i))
					If Not pi Is Nothing Then
						Dim reportable As Reportable() = CType(pi.GetCustomAttributes(GetType(Reportable), False), Reportable())
						If reportable.Length > 0 Then
							If Not reportable(0).AlternateName Is Nothing Then
								altNames(i) = reportable(0).AlternateName
							End If
						Else
							Return ""
						End If

					Else
						Return dataMember
					End If
					If pi.PropertyType.IsGenericType AndAlso (Not pi.PropertyType.IsValueType) Then

						mObj = pi.GetValue(mObj, Nothing)
						mObj = (CType(mObj, IList))(0)
						mType = mObj.GetType()
					End If

					i += 1
				Loop
				Dim s As String = String.Join(".", altNames)
				Return s
			Else
				Return dataMember
			End If
		End Function


		#End Region
	End Class
	#End Region

End Namespace
